sdks/java/core/src/main/java/org/apache/beam/sdk/util/common/Counter.java

```diff
@@ -295,6 +296,56 @@ public AggregationKind getKind() {
   public abstract CounterMean<T> getMean();
 
   /**
+   * CommitState represents whether counters' values have been committed to the backend.
+   *
+   * <p>Runners can use this information to optimize counters updates.
+   * For example, if counters are committed, runners may choose to skip the updates.
+   *
+   * <p>Counters' state transition table:
+   * Action\Current State         COMMITTED        DIRTY        COMMITTING
+   * add()                        DIRTY            DIRTY        DIRTY
+   * committing()                 None             COMMITTING   None
+   * committed()                  None             None         COMMITTED
+   */
+  @VisibleForTesting
+  enum CommitState {
+    COMMITTED,
+    DIRTY,
+    COMMITTING,
+  }
+
+  /**
+   * Returns if the counter contains non-committed aggregate.
+   */
+  public boolean isDirty() {
+    return commitState.get() != CommitState.COMMITTED;
+  }
+
+  /**
+   * Changes the counter from {@code CommitState.DIRTY} to {@code CommitState.COMMITTING}.
+   */
+  public boolean committing() {
+    return commitState.compareAndSet(CommitState.DIRTY, CommitState.COMMITTING);
+  }
+
+  /**
+   * Changes the counter from {@code CommitState.COMMITTING} to {@code CommitState.COMMITTED}.
+   */
+  public boolean committed() {
+    return commitState.compareAndSet(CommitState.COMMITTING, CommitState.COMMITTED);
+  }
+
+  /**
+   * Sets the counter to {@code CommitState.DIRTY}.
+   *
+   * <p>It needs to be called at the beginning of {@link #addValue}, {@link #resetToValue},
```

Comment:
This is making assumptions that the dirty bits cannot be checked in between the addValue() changing the value and before marking the counter dirty. This may not apply to all runners. We need to either (1) clearly document the assumptions that we make about reading counters (eg., when you can/can't trust the dirty bits) or (2) implement this in a way that is safe if a runner decides to look at the dirty bits on a separate thread.